Implement Write trait for MaybeUninit<u8>

SUPERCILEX opened the pull request:
This PR makes it possible to write something like the following:
```rust
let mut buf = [MaybeUninit::<u8>::uninit(); 20];
let written = {
    let slice = buf.as_mut_slice();
    write!(slice, "{}", u64_int).unwrap();
    buf.len() - slice.len()
};
let buf: [u8] = unsafe { mem::transmute(buf[..written]) };
```

rust-highfive commented:
r? @m-ou-se

(rust-highfive has picked a reviewer for you, use r? to override)

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5437035747?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/c7ea1f23c7c42ce7916c644d18d86ed32ec1f0c7/checks/5437035747/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rand v0.7.3
    Checking alloc v0.0.0 (/checkout/library/alloc)
    Checking std v0.0.0 (/checkout/library/std)
    Checking core v0.0.0 (/checkout/library/core)
error[E0282]: type annotations needed for `realstd::vec::Vec<T>`
   |
   |
47 |     let mut buf = Vec::with_capacity(1024);
   |         -------   ^^^^^^^^^^^^^^^^^^ cannot infer type for type parameter `T`
   |         |
   |         consider giving `buf` the explicit type `realstd::vec::Vec<T>`, where the type parameter `T` is specified
error[E0283]: type annotations needed
   --> library/std/src/io/impls/tests.rs:53:24
    |
    |
53  |             let _ = wr.write_all(&src);
    |                     ---^^^^^^^^^------
    |                     |  |
    |                     |  cannot infer type for mutable reference `&mut [_]`
    |                     this method call resolves to `core::result::Result<(), io::error::Error>`
    |
note: multiple `impl`s satisfying `&mut [_]: io::Write` found
    |
    |
334 | impl Write for &mut [u8] {
...
...
424 | impl Write for &mut [mem::MaybeUninit<u8>] {

Some errors have detailed explanations: E0282, E0283.
For more information about an error, try `rustc --explain E0282`.
error: could not compile `std` due to 2 previous errors
```

</details>

SUPERCILEX commented:
Fixed the benchmark and added one for MaybeUninit:

```
test io::impls::tests::bench_write_maybeuninit                    ... bench:          24 ns/iter (+/- 0)
test io::impls::tests::bench_write_slice                          ... bench:          25 ns/iter (+/- 6)
test io::impls::tests::bench_write_vec                            ... bench:          28 ns/iter (+/- 0)
```

SUPERCILEX commented:
Would it make sense to implement Write for `ReadBuf` then?

JohnCSimon commented:
@SUPERCILEX 
Ping from triage: Can you please post the status of this PR?  Maybe ping any commenters for additional feedback?
Thank you.

SUPERCILEX commented:
I'm waiting on a reply to my above comment.

JohnCSimon commented:
@m-ou-se ☝️ ?

bors commented:
:umbrella: The latest upstream changes (presumably #97910) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

JohnCSimon commented:
> I'm waiting on a reply to my above comment.

@m-ou-se - can you comment on this?

SUPERCILEX commented:
This has been implemented in https://github.com/rust-lang/rust/pull/97015. You can now do this:

```rust
let mut buf = [MaybeUninit::<u8>::uninit(); 20];
let mut buf: BorrowedBuf = buf.as_mut_slice().into();
write!(buf.unfilled(), "{}", 42).unwrap();
let buf: &[u8] = buf.filled();
```
